src/rviz/rviz_macros.h

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright (c) 2019, Open Source Robotics Foundation, Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the Willow Garage, Inc. nor the names of its
+ *       contributors may be used to endorse or promote products derived from
+ *       this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef RVIZ_MACROS_H_
+#define RVIZ_MACROS_H_
+
+#include <ros/macros.h>
+
+#ifdef ROS_BUILD_SHARED_LIBS // ros is being built around shared libraries

```

Comment:
Where do these defines (`ROS_BUILD_SHARED_LIBS` and `rviz_EXPORTS`) come from?
I know that cmake can [auto-generate corresponding include files](https://cmake.org/cmake/help/v3.0/module/GenerateExportHeader.html). Why don't you use this mechanism?

Reply:
The primary reason is we saw that many ROS projects have been already following this [ROS Wiki](http://wiki.ros.org/win_ros/Contributing/Dll%20Exports) to define the exports header, and figured we are not going to use another idiom to diverge from this existing one for ROS1 community.

`ROS_BUILD_SHARED_LIBS`: This is defined in `catkin`. See https://github.com/ros/catkin/blob/kinetic-devel/cmake/tools/libraries.cmake#L18

`rviz_EXPORTS`: This is defined by CMake, when the module is built as a shared library. See https://cmake.org/cmake/help/v3.0/prop_tgt/DEFINE_SYMBOL.html